plugins/bufferlines/barbar.nix

```diff
@@ -69,249 +24,567 @@ let
+  deprecateExtraOptions = true;
+  optionsRenamedToSettings = [
+    "animation"
+    "autoHide"
+    "clickable"
+    "focusOnClose"
+    [
+      "hide"
+      "alternate"
+    ]
+    [
+      "hide"
+      "current"
+    ]
+    [
+      "hide"
+      "extensions"
+    ]
+    [
+      "hide"
+      "inactive"
+    ]
+    [
+      "hide"
+      "visible"
+    ]
+    "highlightAlternate"
+    "highlightInactiveFileIcons"
+    "highlightVisible"
+    [
+      "icons"
+      "bufferIndex"
+    ]
+    [
+      "icons"
+      "bufferNumber"
+    ]
+    [
+      "icons"
+      "button"
+    ]
+    [
+      "icons"
+      "filetype"
+      "customColors"
+    ]
+    [
+      "icons"
+      "separator"
+      "left"
+    ]
+    [
+      "icons"
+      "separator"
+      "right"
+    ]
+    "insertAtStart"
+    "insertAtEnd"
+    "maximumPadding"
+    "minimumPadding"
+    "maximumLength"
+    "semanticLetters"
+    "letters"
+    "sidebarFiletypes"
+    "noNameTitle"
+    "tabpages"
+  ];
+  imports =
+    let
+      basePluginPath = [
+        "plugins"
+        "barbar"
+      ];
+      settingsPath = basePluginPath ++ [ "settings" ];
+    in
+    [
+      (mkRemovedOptionModule (
+        basePluginPath
+        ++ [
+          "keymaps"
+          "silent"
+        ]
+      ) "Keymaps will be silent anyways. This option has always been useless.")
+      (mkRenamedOptionModule (basePluginPath ++ [ "excludeFileTypes" ]) (
+        settingsPath ++ [ "exclude_ft" ]
+      ))
+      (mkRenamedOptionModule (basePluginPath ++ [ "excludeFileNames" ]) (
+        settingsPath ++ [ "exclude_name" ]
+      ))
+      (mkRemovedOptionModule (
+        basePluginPath
+        ++ [
+          "icons"
+          "diagnostics"
+        ]
+      ) "Use `settings.icons.diagnostics` instead, but pay attention as the keys have changed.")
+      (mkRenamedOptionModule
+        (
+          basePluginPath
+          ++ [
+            "icons"
+            "filetype"
+            "enable"
+          ]
+        )
+        (
+          settingsPath
+          ++ [
+            "icons"
+            "filetype"
+            "enabled"
+          ]
+        )
+      )
+    ]
+    ++ (map
+      (
+        name:
+        mkRemovedOptionModule (
+          basePluginPath
+          ++ [
+            "icons"
+            name
+          ]
+        ) "Use `settings.icons.${name}` instead, but you should now use the real `snake_case` key names."
+      )
+      [
+        "alternate"
+        "current"
+        "inactive"
+        "modified"
+        "pinned"
+        "visible"
+      ]
+    );
+
+  extraOptions = {
+    keymaps = mapAttrs (
+      optionName: funcName:
+      helpers.mkNullOrOption (helpers.keymaps.mkMapOptionSubmodule {
+        defaults = {
+          mode = "n";
+          action = "<Cmd>Buffer${funcName}<CR>";
+        };
+        hasKey = true;
+        hasAction = true;
+      }) "Keymap for function Buffer${funcName}"
+    ) keymapsActions;
```

Comment:
One limitation of `keymaps` being an attrset is that you can only define _one_ keymap per action.

At least one other plugin uses a list of submodules, with the `action` type being an enum of `optionNames` (`attrNames keymapsActions`): https://github.com/nix-community/nixvim/blob/f4ce7dad0e9ba16692783a70ed7f5b79e41d6c58/plugins/utils/improved-search.nix#L38-L54

That said _most_ plugin keymaps so far are using attrsets like this one.

Reply:
Yes, I would keep it like this for this plugin. Again, the user can always fall back to the top-level `keymaps` option.